File: leetcode/medium/TopKFrequentElements.py
from typing import List
import collections


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """
        Finds the k most frequent elements in an array of integers.

        This solution uses a bucket sort-like approach to achieve O(N) time complexity,
        which is more efficient than a sorting-based O(N log N) solution.

        Algorithm:
        1.  Count the frequency of each number in the input array using a hash map
            (or `collections.Counter`). This takes O(N) time.
        2.  Create a list of "buckets", where the index of each bucket corresponds to a
            frequency. `buckets[i]` will store a list of all numbers that appear `i` times.
            The size of this list will be `len(nums) + 1`.
        3.  Iterate through the frequency map and place each number into its corresponding
            frequency bucket. This takes O(U) time, where U is the number of unique elements.
        4.  Iterate through the buckets in reverse order (from highest frequency to lowest).
            Collect the numbers from the buckets until `k` elements have been gathered.

        Time Complexity: O(N) - The frequency counting is O(N), populating the buckets is O(U),
        and gathering the result is O(U) in the worst case. Since U <= N, the overall
        complexity is dominated by the initial counting step, making it O(N).
        Space Complexity: O(N) - For the frequency map and the buckets array.
        """
        if k == 0:
            return []

        # Step 1: Count the frequency of each number. O(N)
        # collections.Counter is a highly efficient and Pythonic way to do this.
        freq_map = collections.Counter(nums)

        # Step 2: Create buckets to group numbers by frequency.
        # The index is the frequency, and the value is a list of numbers with that frequency.
        buckets = [[] for _ in range(len(nums) + 1)]
        for num, freq in freq_map.items():
            buckets[freq].append(num)

        # Step 3: Gather the top k elements by iterating through buckets from high to low frequency.
        result = []
        for i in range(len(buckets) - 1, 0, -1):
            result.extend(buckets[i])
            if len(result) >= k:
                # We only need the top k, so we can slice and return.
                return result[:k]
        return result # Should only be reached if k > number of unique elements

File: leetcode/medium/TopKFrequentElements.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
// java.util.Map.Entry is not strictly needed for the enhanced for loop over entrySet()
// We also don't need java.util.Arrays for the logic, but it's often imported.

/**
 * Solves the "Top K Frequent Elements" problem using the Bucket Sort technique.
 * This method achieves an optimal time complexity of O(N), where N is the number
 * of elements in the input array.
 */
class Solution {
    /**
     * Finds the k most frequently occurring elements in the array nums.
     *
     * @param nums The input array of integers.
     * @param k The number of most frequent elements to find.
     * @return An array containing the k most frequent elements.
     */
    public int[] topKFrequent(int[] nums, int k) {
        // 1. Calculate the frequency of each number (Time: O(N))
        // The key is the number, and the value is its frequency (count).
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : nums) {
            // Use getOrDefault to simplify the frequency counting logic.
            frequencyMap.put(num, frequencyMap.getOrDefault(num, 0) + 1);
        }

        // 2. Initialize a Bucket Array (Time: O(N))
        // The index represents the frequency, and the list at that index
        // stores the numbers that have that exact frequency.
        // The maximum possible frequency is nums.length.
        // We use an array of Lists, where the size is N + 1 (for frequencies 0 to N).
        // Using an array of List<Integer> and initializing elements as null is also an option,
        // but initializing all lists upfront simplifies step 3's logic.
        @SuppressWarnings("unchecked")
        List<Integer>[] buckets = new List[nums.length + 1];
        for (int i = 0; i < buckets.length; i++) {
            buckets[i] = new ArrayList<>();
        }

        // 3. Place numbers into the correct frequency bucket (Time: O(M), where M is the number of unique elements, M <= N)
        for (Map.Entry<Integer, Integer> entry : frequencyMap.entrySet()) {
            int number = entry.getKey();
            int freq = entry.getValue();
            // Place the number into the bucket corresponding to its frequency.
            buckets[freq].add(number);
        }

        // 4. Collect the results from the buckets, starting from the highest frequency (Time: O(N))
        List<Integer> resultList = new ArrayList<>();
        // Iterate backwards from the largest possible frequency (nums.length) down to 1.
        for (int i = nums.length; i >= 1; i--) {
            // Check if the current frequency bucket is non-empty.
            if (!buckets[i].isEmpty()) {
                // Add all numbers in the current bucket to the result.
                for (int number : buckets[i]) {
                    if (k > 0) {
                        resultList.add(number);
                        k--;
                    } else {
                        // Optimization: if k reaches 0, we have found the top k elements.
                        break;
                    }
                }
            }
            // If k reaches 0 after processing a bucket, we stop the outer loop.
            if (k == 0) {
                break;
            }
        }

        // 5. Convert the List<Integer> to int[] (Time: O(K))
        // Instead of using streams, a simple loop is generally faster and avoids boxing/unboxing overhead.
        int[] finalResult = new int[resultList.size()];
        for (int i = 0; i < resultList.size(); i++) {
            finalResult[i] = resultList.get(i);
        }

        return finalResult;
    }
}